opencensus/exporters/trace/zipkin/zipkin_exporter.cc

```diff
@@ -0,0 +1,342 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "zipkin_exporter.h"
+
+#include <curl/curl.h>
+#include <rapidjson/prettywriter.h>
+#include <rapidjson/stringbuffer.h>
+#include "absl/strings/str_cat.h"
+
+namespace opencensus {
+namespace exporters {
+namespace trace {
+
+namespace {
+
+constexpr const char *kZipkinLibname = "zipkin";
+constexpr const char *kZkipkinVersion = "2.0";
+
+void SerializeJson(rapidjson::Writer<rapidjson::StringBuffer> *writer,
+                   const ::opencensus::trace::exporter::SpanData &span) {
+  writer->StartObject();
+
+  writer->Key("name");
+  writer->String(std::string(span.name()));
+
+  writer->Key("traceId");
+  writer->String(span.context().trace_id().ToHex());
+
+  writer->Key("id");
+  writer->String(span.context().span_id().ToHex());
+
+  writer->Key("parentId");
+  writer->String(span.parent_span_id().ToHex());
+
+  writer->Key("annotations");
+  writer->StartArray();
+  for (const auto &annotation : span.annotations().events()) {
+    writer->StartObject();
+    writer->Key("timestamp");
+    writer->Int64(absl::ToUnixMicros(annotation.timestamp()));
+    writer->Key("value");
+    writer->String(annotation.event().DebugString());
+    writer->EndObject();
+  }
+  writer->EndArray(span.annotations().events().size());
+
+  writer->Key("binaryAnnotations");
+  writer->StartArray();
+  for (const auto &attribute : span.attributes()) {
+    writer->StartObject();
+    writer->Key("key");
+    writer->String(attribute.first);
+    writer->Key("value");
+    switch (attribute.second.type()) {
+      case ::opencensus::trace::AttributeValueRef::Type::kString:
+        writer->String(attribute.second.string_value());
+        writer->Key("type");
+        writer->String("STRING");
+        break;
+      case ::opencensus::trace::AttributeValueRef::Type::kBool:
+        writer->Bool(attribute.second.bool_value());
+        writer->Key("type");
+        writer->String("BOOL");
+        break;
+      case ::opencensus::trace::AttributeValueRef::Type::kInt:
+        writer->Int64(attribute.second.int_value());
+        writer->Key("type");
+        writer->String("I64");
+        break;
+      default:
+        break;
+    }
+    writer->EndObject();
+  }
+  writer->EndArray(span.attributes().size());
+
+  writer->Key("timestamp");
+  writer->Int64(absl::ToUnixMicros(span.start_time()));
+
+  writer->Key("duration");
+  writer->Int64(absl::ToInt64Microseconds(span.end_time() - span.start_time()));
+
```

Comment:
Links? MessageEvents?

Reply:
I didn't see an equivalent of those in the Zipkin format.